File: app/src/main/java/org/scp/gymlog/ui/main/muscles/MusclesRecyclerViewAdapter.java
package org.scp.gymlog.ui.main.muscles;

import android.content.Context;
import android.content.res.ColorStateList;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.databinding.ListElementFragmentMuscleBinding;
import org.scp.gymlog.model.Muscle;
import org.scp.gymlog.util.Data;

import java.util.List;
import java.util.function.Consumer;

public class MusclesRecyclerViewAdapter extends RecyclerView.Adapter<MusclesRecyclerViewAdapter.ViewHolder> {

	private final List<Muscle> muscles;
	private final Consumer<Muscle> onClickElementListener;

	public MusclesRecyclerViewAdapter(@NonNull Consumer<Muscle> onClickElementListener) {
		this.onClickElementListener = onClickElementListener;
		muscles = Data.getInstance().getMuscles();
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		return new ViewHolder(
				ListElementFragmentMuscleBinding.inflate(
						LayoutInflater.from(parent.getContext()), parent, false
				)
		);
	}

	@Override
	public void onBindViewHolder(final ViewHolder holder, int position) {
		Context context = holder.itemView.getContext();

		holder.muscle = muscles.get(position);
		holder.mContentView.setText(holder.muscle.getText());

		holder.mImageView.setImageResource(holder.muscle.getIcon());
		holder.mImageView.setImageTintList(ColorStateList.valueOf(
				context.getResources().getColor(holder.muscle.getColor(), null)
		));

		holder.mIndicator.setBackgroundResource(holder.muscle.getColor());
	}

	@Override
	public int getItemCount() {
		return muscles.size();
	}

	public class ViewHolder extends RecyclerView.ViewHolder {
		public Muscle muscle;
		public final TextView mContentView;
		public final ImageView mImageView;
		public final View mIndicator;

		public ViewHolder(ListElementFragmentMuscleBinding binding) {
			super(binding.getRoot());
			mContentView = binding.content;
			mImageView = binding.image;
			mIndicator = binding.indicator;

			itemView.setOnClickListener(a-> onClickElementListener.accept(muscle));
		}

		@Override
		public String toString() {
			return super.toString() + " '" + mContentView.getText() + "'";
		}
	}
}

File: app/src/main/java/org/scp/gymlog/ui/main/muscles/MusclesRecyclerViewAdapter.kt
package org.scp.gymlog.ui.main.muscles

import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.databinding.ListElementFragmentMuscleBinding
import org.scp.gymlog.model.Muscle
import org.scp.gymlog.util.Data
import java.util.function.Consumer

class MusclesRecyclerViewAdapter(
	private val onClickElementListener: Consumer<Muscle>
) : RecyclerView.Adapter<MusclesRecyclerViewAdapter.ViewHolder>() {

	private val muscles: List<Muscle>

	init {
		muscles = Data.muscles
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		return ViewHolder(
			ListElementFragmentMuscleBinding.inflate(
				LayoutInflater.from(parent.context), parent, false
			)
		)
	}

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		val context = holder.itemView.context
		val muscle = muscles[position].also { holder.muscle = it }

		holder.mContentView.setText(muscle.text)
		holder.mImageView.setImageResource(muscle.icon)
		holder.mImageView.imageTintList = ColorStateList.valueOf(
			context.resources.getColor(muscle.color, null)
		)
		holder.mIndicator.setBackgroundResource(muscle.color)
	}

	override fun getItemCount(): Int {
		return muscles.size
	}

	inner class ViewHolder(binding: ListElementFragmentMuscleBinding) :
		RecyclerView.ViewHolder(binding.root) {

		var muscle: Muscle? = null
		val mContentView: TextView = binding.content
		val mImageView: ImageView = binding.image
		val mIndicator: View = binding.indicator

		init {
			itemView.setOnClickListener { onClickElementListener.accept(muscle!!) }
		}

		override fun toString(): String {
			return super.toString() + " '" + mContentView.text + "'"
		}
	}
}
